File: kotlin/java-basic-example-vs-kotlin/src/main/java/org/example/ex10/Exam10.java
package org.example.ex10;

import java.util.Optional;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * @author : Rene
 * @since : 2023/10/22
 */
public class Exam10 {

	public Exam10(ExamUser examUser){
		if (examUser!=null && examUser.getName() != null){
			if (!examUser.getName().isBlank()){
				// 로직
			}
		}

		// 혹은

		var optionalUser = Optional.ofNullable(examUser);

		optionalUser.ifPresent(it ->{
			Optional.ofNullable(examUser.getName()).ifPresent(name->{
				if(!name.isBlank()){
					// 로직
				}
			});
		});

		// 이러한 방어 코드를 작성하게 됨

	}
	public static void main(String[] args) {
		new Exam10(new ExamUser());
		new Exam10(new ExamUser("aa"));
	}
}



@Data
@NoArgsConstructor
@AllArgsConstructor
class ExamUser{
	private String name;
}




// notblank를 사용하려고 이렇게 util 함수를 만들기도 한다.
class StringUtils{
	public static boolean notBlank(String value){
		return !value.isBlank();
	}
}


class ObjectUtils{
	public static boolean isNotNull(Object object){
		return object != null;
	}
}


// 이러한 코드를 코틀린에서는 ?

File: kotlin/kotlin-basic-example/src/main/kotlin/ex10/Exam10.kt
package ex10

/**
 * @author        : Rene
 * @since         : 2023/10/22
 */
fun main(){
    val user = ExamUser(name = "abcd");
    exam10(user)
}



fun exam10(examUser: ExamUser?){
    examUser?.let {
        it.name?.let{name ->
            println(name)
        }
    }


    examUser?.let{
        if(it.name.isNullOrEmpty()){
            // 로직
        }

        // 뒤에 만든 함수를 마치 스트링 클래스에 생긴 메서드처럼 동작을 함 -> 내가 원하는 특정 클래스를 이미 존재하는 클래스에 추가가 가능함
        if(it.name.isNotNullOrBlank()){

        }



        if (examUser.isNotNull() && examUser.name.isNotNullOrBlank()){
            // 로직
        }

    }
}
data class ExamUser(
    var name: String ?=null
)

// 코틀린의 확장 함수 -> extension function
fun String.isNotNullOrBlank(): Boolean{
    return !this.isNotNullOrBlank();
}



// 코틀린에서 모든 Object의 조상은 Any

fun Any?.isNotNull():Boolean{
    return this!=null;
}


// 중요한 점은 확장 함수 -> 자주 사용하는 함수들을 만들 수도 있고 이미 만들어진 것을 정의할 수 있음
